addon/-private/system/relationships/relationship-payloads.js

```diff
@@ -0,0 +1,283 @@
+  }
+
+  _unload(id, payloads, inversePayloads) {
+    if (!this._inverseLoaded(payloads[id])) {
+      this._removeInverse(id, payloads[id], inversePayloads);
+      delete payloads[id];
+    }
+  }
+
+  _flushPending() {
+    if (this._pendingPayloads.length === 0) { return; }
+
+    let payloadsToBeProcessed = this._pendingPayloads.splice(0, this._pendingPayloads.length);
+    for (let i=0; i<payloadsToBeProcessed.length; ++i) {
+      let modelName = payloadsToBeProcessed[i][0];
+      let id = payloadsToBeProcessed[i][1];
+      let relationshipName = payloadsToBeProcessed[i][2];
+      let relationshipData = payloadsToBeProcessed[i][3];
+
+      // TODO: maybe delay this slightly?
+      let inverseRelationshipData = {
+        data: {
+          id: id,
+          type: modelName
+        }
+      }
+
+      // start flushing this individual payload.  The logic is the same whether
+      // it's for the left hand side of the relationship or the right hand side,
+      // except the role of primary and inverse idToPayloads is reversed
+      //
+      let previousPayload;
+      let idToPayloads;
+      let inverseIdToPayloads;
+      let inverseIsMany;
+      if (this._isLHS(modelName, relationshipName)) {
+        previousPayload = this._lhsPayloads[id];
+        idToPayloads = this._lhsPayloads;
+        inverseIdToPayloads = this._rhsPayloads;
+        inverseIsMany = this._rhsRelationshipIsMany;
+      } else {
+        assert(`${modelName}:${relationshipName} is not either side of this relationship, ${this._lhsModelName}:${this._lhsRelationshipName}<->${this._rhsModelName}:${this._rhsRelationshipName}`, this._isRHS(modelName, relationshipName));
+        previousPayload = this._rhsPayloads[id];
+        idToPayloads = this._rhsPayloads;
+        inverseIdToPayloads = this._lhsPayloads;
+        inverseIsMany = this._lhsRelationshipIsMany;
+      }
+
+      // actually flush this individual payload
+      //
+      this._removeInverse(id, previousPayload, inverseIdToPayloads);
+      idToPayloads[id] = relationshipData;
+      this._populateInverse(relationshipData, inverseRelationshipData, inverseIdToPayloads, inverseIsMany);
+    }
+  }
+
+  _inverseLoaded(payload) {
+    let data = payload && payload.data;
+    if (!data) { return false; }
+
+    if (Array.isArray(data)) {
+      for (let i=0; i<data.length; ++i) {
+        if (hasRecordForId(this._store, data[i].type, data[i].id)) {
+          return true;
+        }
+      }
+      return false;
+    } else {
+      return hasRecordForId(this._store, data.type, data.id);
+    }
+  }
+
+  _populateInverse(relationshipData, inversePayload, inverseIdToPayloads, inverseIsMany) {
+    if (!relationshipData.data) {
+      // This id doesn't have an inverse, eg a belongsTo with a payload
+      //-{ data: null }, so there's nothing to populate
+      return;
+    }
+
+    if (Array.isArray(relationshipData.data)) {
+      for (let i=0; i<relationshipData.data.length; ++i) {
```

Comment:
https://jsperf.com/map-fn-vs-map-arrow-vs-fori/1

unless i've messed something up with the benchmarks, fori looks to be about 2x faster than map